### PR TITLE
Added sniff for deprecated model methods. #187

### PR DESCRIPTION
**Pull request by milindsingh**
Added sniff for #187 to discourage the use of Resource class from Model class.

Trigger Methods: 

- getResource()->load($object)
- getResource()->save($object)
- getResource()->delete($object)

Message: `Possible use of the deprecated model method 'getResource()' to 'save' the data detected.`

### CODE REVIEW

File: Magento2/Sniffs/Methods/DeprecatedModelMethodSniff.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    protected $severity = 0;
+
+    /**
+     * @inheritdoc
+     */
+    public function register()
+    {
+        return [
+            T_OBJECT_OPERATOR,
+            T_DOUBLE_COLON
```
Comment (lenaorobei) [inline Magento2/Sniffs/Methods/DeprecatedModelMethodSniff.php:52]: Don't think we need `T_DOUBLE_COLON` here.

File: Magento2/Sniffs/Methods/DeprecatedModelMethodSniff.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        ];
+    }
+    /**
+     * @inheritdoc
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $tokens = $phpcsFile->getTokens();
+        $methodPosition = $phpcsFile->findNext(T_STRING, $stackPtr + 1);
```
Comment (lenaorobei) [inline Magento2/Sniffs/Methods/DeprecatedModelMethodSniff.php:61]: Please limit scope for search.

File: Magento2/Sniffs/Methods/DeprecatedModelMethodSniff.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+        if ($methodPosition !== false &&
+            in_array($tokens[$methodPosition]['content'], $this->methods)
+        ) {
+            $resourcePosition = $phpcsFile->findPrevious([T_STRING, T_VARIABLE], $stackPtr - 1);
```
Comment (lenaorobei) [inline Magento2/Sniffs/Methods/DeprecatedModelMethodSniff.php:66]: I propose to change the logic:
1. Search for `getResource`
2. After that check if listed methods are called.

File: Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<?php
+$model->getResource()->save();
```
Comment (lenaorobei) [inline Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc:2]: Please add positive cases as well.

File: Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+$model->getResource()->load();
+$model->getResource()->delete();
```
Comment (lenaorobei) [inline Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc:4]: Missing new line.

File: Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc (hunk 4 above)
Comment (milindsingh) [reply Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc]: Thanks @lenaorobei 
Can you please suggest a negative case that I can add?

File: Magento2/Sniffs/Methods/DeprecatedModelMethodSniff.php (hunk 2 above)
Comment (milindsingh) [reply Magento2/Sniffs/Methods/DeprecatedModelMethodSniff.php]: Thanks @lenaorobei 

I have updated to following:
`$resourcePosition = $phpcsFile->findNext(T_STRING, $stackPtr + 1, $endOfStatement, false, self::RESOURCE_METHOD);`

File: Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc (hunk 4 above)
Comment (ihor-sviziev) [reply Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc]: @milindsingh,
What about this?
```php
$model->getResource()->myCustomMethod();
$model->myCustomMethod();
```

File: Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+$model->getResource()->save($model);
+
+$model->getResource()->load($model, $id);
+
+$model->getResource()->delete($model);
```
Comment (lenaorobei) [inline Magento2/Tests/Methods/DeprecatedModelMethodUnitTest.inc:7]: Please add here something like.

```
$model->anotherMethodWithResource()->save($model);
```

We need have cases here to make sure check does not cause false positive findings.

**Comment by milindsingh** [comment]
Thanks @ihor-sviziev 
@lenaorobei I think its good to merge now.

**Comment by lenaorobei** [comment]
Great! Thanks @ihor-sviziev. Let me run this test again Magento codebase and make sure it does not produce false positive findings.

**Pull request merged**